src/platform/ESP32/CHIPMem-Platform.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements heap memory allocation APIs for CHIP. These functions are platform
+ *      specific and might be C Standard Library heap functions re-direction in most of cases.
+ *
+ */
+
+#include <lib/support/CHIPMem.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#include <atomic>
+#include <cstdio>
+#include <cstring>
+#include <esp_heap_caps.h>
+#include <stdlib.h>
+
+namespace chip {
+namespace Platform {
+
+#define VERIFY_INITIALIZED() VerifyInitialized(__func__)
+
+static std::atomic_int memoryInitialized{ 0 };
+
+static void VerifyInitialized(const char * func)
+{
+    if (!memoryInitialized)
+    {
+        ChipLogError(DeviceLayer, "ABORT: chip::Platform::%s() called before chip::Platform::MemoryInit()", func);
+        abort();
+    }
+}
+
+CHIP_ERROR MemoryAllocatorInit(void * buf, size_t bufSize)
+{
+    if (memoryInitialized++ > 0)
+    {
+        ChipLogError(DeviceLayer, "ABORT: chip::Platform::MemoryInit() called twice.");
```

Comment:
is there a reason this would be an error? If allocator init is a noop, it seems not useful to track these counts and error out.

Maybe a comment explaining why we are so string would help.

Reply:
Yes, changed the init and shutdown function.